airbyte-cdk/python/airbyte_cdk/sources/streams/core.py

```diff
@@ -328,6 +381,43 @@ def get_updated_state(
         """
         return {}
 
+    def _get_checkpoint_reader(
+        self,
+        logger: logging.Logger,
+        cursor_field: Optional[List[str]],
+        sync_mode: SyncMode,
+        stream_state: MutableMapping[str, Any],
+    ) -> CheckpointReader:
+        checkpoint_mode = self._checkpoint_mode
+        if checkpoint_mode == CheckpointMode.RESUMABLE_FULL_REFRESH:
+            return ResumableFullRefreshCheckpointReader(stream_state=stream_state)
+        else:
+            slices = self.stream_slices(
+                cursor_field=cursor_field,
+                sync_mode=sync_mode,  # todo: change this interface to no longer rely on sync_mode for behavior
+                stream_state=stream_state,
+            )
+            # Because of poor foresight, we wrote the default Stream.stream_slices() method to return [None] which is confusing and
+            # now normalized this behavior for connector developers. Now some connectors also return [None]. This is objectively
+            # misleading and a more ideal interface is [{}] to indicate we still want to iterate over one slice, but with no
+            # specific slice values. None is bad, and now I feel bad that I have to write this hack.
+            if slices == [None]:
+                slices = [{}]
+            logger.debug(f"Processing stream slices for {self.name}", extra={"stream_slices": slices})
```

Comment:
fun fact: slices is a generator so it won't even be logged properly

Reply:
my god this is awful. I'm just getting rid of this